cmd/transcriber/call/tracks.go

```diff
@@ -304,13 +306,77 @@ func (t *Transcriber) transcribeTrack(ctx trackContext) (transcribe.TrackTranscr
 		return trackTr, 0, fmt.Errorf("failed to decode audio samples: %w", err)
 	}
 
+	slog.Debug("decoding done", slog.Any("samplesLen", len(samples)))
+
 	transcriber, err := t.newTrackTranscriber()
 	if err != nil {
 		return trackTr, 0, fmt.Errorf("failed to create track transcriber: %w", err)
 	}
 
-	var totalDur time.Duration
+	sd, err := speech.NewDetector(speech.DetectorConfig{
+		ModelPath:            filepath.Join(getModelsDir(), "silero_vad.onnx"),
+		SampleRate:           trackOutAudioRate,
+		WindowSize:           1536,
+		Threshold:            0.5,
+		MinSilenceDurationMs: 1000,
+		SpeechPadMs:          100,
+	})
+	if err != nil {
+		return trackTr, 0, fmt.Errorf("failed to ceate speech detector: %w", err)
+	}
+	defer func() {
+		if err := sd.Destroy(); err != nil {
+			slog.Error("failed to destroy speech detector", slog.String("err", err.Error()), slog.String("trackID", ctx.trackID))
+		}
+	}()
+
+	// Before transcribing, we feed the samples to a speech detector and adjust
+	// the timestamps in accordance to when the speech begins/ends. This is
+	// to account for any potential silence that Whisper wouldn't recognize with
+	// much accuracy.
+	// TODO: consider deprecating this logic if we get accurate word level timestamps
+	// (https://github.com/ggerganov/whisper.cpp/issues/375).
+
+	var speechSamples []trackTimedSamples
 	for _, ts := range samples {
+		segments, err := sd.Detect(ts.pcm)
+		if err != nil {
+			slog.Error("failed to detect speech",
+				slog.String("err", err.Error()),
+				slog.String("trackID", ctx.trackID))
+			continue
+		}
+		slog.Debug("speech detection done", slog.Any("segments", segments))
+
+		for _, seg := range segments {
+			startSampleOff := int(seg.SpeechStartAt * trackOutAudioRate)
+			endSampleOff := int(seg.SpeechEndAt * trackOutAudioRate)
```

Comment:
One request -- can you add some comments for future us about what the units are for `seg.SpeechStartAt` and `startSampleOff`? (That will help in the future when I'm reviewing and remembering what the numbers mean.)

Reply:
Sure but the sample offset is just what it says. It's an index to the first sample where the speech starts (+- some padding), there's no unit really.